sound/soc/sof/intel/hda.c

```diff
@@ -1384,16 +1383,19 @@ void hda_machine_select(struct snd_sof_dev *sdev)
 	/*
 	 * If I2S fails, try SoundWire
 	 */
-	hda_sdw_machine_select(sdev);
+	if (!mach)
+		mach = hda_sdw_machine_select(sdev);
 
 	/*
 	 * Choose HDA generic machine driver if mach is NULL.
 	 * Otherwise, set certain mach params.
 	 */
-	hda_generic_machine_select(sdev);
+	mach = hda_generic_machine_select(sdev, mach);
```

Comment:
it's just awful to have both a modifiable parameter and a return value. I have a visceral reaction against this.

Reply:
struct snd_soc_acpi_mach *mach;

Nothing prevents the hda_generic_machine_select from modifying the fields of that structure and return a different pointer. It's just confusing.

Reply:
@plbossart it was the same before that but hidden and dropping `const`... We can make `hda_generic_machine_select(sdev, &mach);` but that doesn't seem better to me. The current code is already doing that: it can replace the `pdata->machine` pointer, and it modifies its contents:
```
if (pdata->machine) {
	mach_params = (struct snd_soc_acpi_mach_params *)
		&pdata->machine->mach_params;
	mach_params->codec_mask = bus->codec_mask;
	mach_params->common_hdmi_codec_drv = hda_codec_use_common_hdmi;
}
```

Reply:
Your missing my point @lyakh: using the same variable for the return and the argument of a function gives me a migraine. Pick one, not both.

Reply:
`x = max(x, y);` seems to be a very commonly used pattern, although it's actually (theoretically) wasting CPU cycles, or relying on the compiler to optimise the redundant assignment away. But ok, I'll change that to `hda_generic_machine_select(sdev, &mach);`

Reply:
for scalars that's fine indeed. It's when you handle structures that things can be odd and difficult to review. Thanks for doing the change, I appreciate it.